### ISSUE TITLE
Overloads returning type predicate types are always incompatible with implementation

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#4180]
Version 1.6.0-dev.20150805

```TypeScript
function f(x: string): x is string;
function f(x: number): x is number;
function f(x: any): boolean {
    return true;
}
```

Currently yields

```
blah.ts(1,10): error TS2394: Overload signature is not compatible with function implementa
tion.
```

I think this should be fixed in time for 1.6; being able to do this would enable some useful patterns.

**Comment by DanielRosenwasser** [comment]
Specific use case found with @roganov [here](https://gitter.im/Microsoft/TypeScript?at=55c2f9e77a6037e67c59c10c).

```TypeScript
namespace ActionType {
    export type Foo = string & { _fooTag: any };
    export const Foo = <Foo>"Foo";

    export type Bar = string & { _barTag: any };
    export const Bar = <Bar>"Bar";
}

type ActionType = ActionType.Foo |
                  ActionType.Bar;


interface BaseAction {
    type: ActionType;
}

interface FooAction extends BaseAction {
    type: ActionType.Foo;
    foo: number;
}

interface BarAction extends BaseAction {
    type: ActionType.Bar;
    bar: string;
}

function isActionType(action: BaseAction, type: ActionType.Foo): action is FooAction;
function isActionType(action: BaseAction, type: ActionType.Bar): action is BarAction;
function isActionType(action: BaseAction, type: ActionType): boolean {
    return action.type === type;
}

let handleAction = (action: BaseAction) => {
    if (isActionType(action, ActionType.Foo)) {
        let foo = action.foo; // OK
        let bar = action.bar; // type error
    }
    else if (isActionType(action, ActionType.Bar)) {
        let bar = action.bar; // OK
    }
}
```

**Comment by tinganho** [comment]
See https://github.com/Microsoft/TypeScript/pull/3262#issuecomment-108991303.

**Comment by DanielRosenwasser** [comment]
Gotcha - but there was no formal issue filed for that, correct?

**Comment by tinganho** [comment]
Correct, at least non that I knew of.

**Comment by DanielRosenwasser** [comment]
@JsonFreeman in the comment that @tinganho linked to, you mentioned that this is a problem that "we've known about". What problem were you referring to specifically?

**Comment by JsonFreeman** [comment]
I was referring to #943.

**Comment by DanielRosenwasser** [comment]
Right, I was sure we had an issue referring to that. Glad I'm not going nuts :smile:. @mhegazy.

**Comment by DanielRosenwasser** [comment]
This should be fixed for TypeScript 1.8.
